### PR TITLE
Adding on-chain libraries TEP

### PR DESCRIPTION
**Pull request by kehiy**
I wrote this proposal to introduce a way to make a blockchain more scaleable and lightweight.
Maybe some part of this proposal need to be changed and it needs more discussion to have final changes and versions.

> Note: I Used empty TEP number to be considered.

### CODE REVIEW

**Comment by kehiy** [comment]
@tolya-yanot @EmelyanenkoK @hacker-volodya 

Hi, can you have a review guys? should I publish this in any specific place?

**Comment by ProgramCrafter** [comment]
I shall note that this is already implemented, only that it's currently inconvenient to call specific functions out of library; thus, the changes would only be within FunC compiler (though, groups of functions are more suited to Tact).

---

> 1. The most used TON library (stdlib.fc) can be deployed on-chain which is used in most of the contracts (including wallets).

It would *increase* fees, as stdlib is already written in TVM assembly (and something is even emitted by FunC compiler).

**Comment by kehiy** [comment]
@ProgramCrafter, thank you for your response. so, you are saying that at the current moment, there is no duplicated stdlib code on TON blockchain? or I misunderstood it?

File: text/0000-on-chain-libraries.md
Comment (ahmadrafiee331) [inline text/0000-on-chain-libraries.md:1]: UQBhtq4l0mRuN92W5Ak1ILjIrIswuLPUpVKSUC4jaST0qgE5

**Review by ahmadrafiee331 — APPROVED** [review]
UQBhtq4l0mRuN92W5Ak1ILjIrIswuLPUpVKSUC4jaST0qgE5

**Comment by EmelyanenkoK** [comment]
> @ProgramCrafter, thank you for your response. so, you are saying that at the current moment, there is no duplicated stdlib code on TON blockchain? or I misunderstood it?

Stdlib.fc generally doesn't actually contain new code, it consists of useful funC aliases to asm instructions.

Library cells as reusable and cheap to storage references to constant "chunks" of data (it also may be treated as chunks of code) is already  implemented, you can find it in the docs: https://docs.ton.org/develop/data-formats/library-cells

**Pull request closed without merging**